File: apps/atlas/test/heading_service_test.dart
// ============================================================
// As Above, So Below. As Within, So Without.
// The Future Dictates the Past and the Past is Always Present.
// ============================================================

import 'package:flutter_test/flutter_test.dart';

import 'package:atlas/location/heading_service.dart';

import 'fake_heading_source.dart';

HeadingService supportedService(FakeHeadingSource fake) {
  final service = HeadingService(source: fake);
  addTearDown(service.dispose);
  return service;
}

void main() {
  test('initial state is idle with unknown accuracy', () {
    final service = HeadingService(source: FakeHeadingSource());
    expect(service.supported, isNull);
    expect(service.unsupported, isFalse);
    expect(service.latest, isNull);
    expect(service.displayDeg, isNull);
    expect(service.accuracy, HeadingAccuracy.unknown);
    expect(service.active, isFalse);
    expect(service.dimmed, isTrue);
  });

  test('unsupported sensor is explicit and never subscribes', () async {
    final fake = FakeHeadingSource()..supportedResult = false;
    final service = supportedService(fake);
    await service.ensureStarted();
    expect(service.unsupported, isTrue);
    expect(service.active, isFalse);
    expect(fake.samplesCalls, 0);
  });

  test('start subscribes exactly once', () async {
    final fake = FakeHeadingSource();
    final service = supportedService(fake);
    await service.start();
    await service.start();
    await service.start();
    expect(service.supported, isTrue);
    expect(service.active, isTrue);
    expect(fake.samplesCalls, 1);
  });

  test('ingested sample prefers true north', () async {
    final fake = FakeHeadingSource();
    final service = supportedService(fake);
    await service.start();
    fake.emit(testSample(magnetic: 90.0, trueNorth: 95.0));
    await Future<void>.delayed(Duration.zero);
    expect(service.displayDeg, 95.0);
    expect(service.frameLabel, 'TRUE');
    expect(service.accuracy, HeadingAccuracy.high);
    expect(service.dimmed, isFalse);
  });

  test('missing true north falls back to magnetic labeled MAG', () async {
    final fake = FakeHeadingSource();
    final service = supportedService(fake);
    await service.start();
    fake.emit(testSample(magnetic: 90.0, trueNorth: null));
    await Future<void>.delayed(Duration.zero);
    expect(service.displayDeg, 90.0);
    expect(service.frameLabel, 'MAG');
  });

  test('low and unreliable accuracy dim the compass', () async {
    final fake = FakeHeadingSource();
    final service = supportedService(fake);
    await service.start();
    fake.emit(testSample(accuracy: HeadingAccuracy.low));
    await Future<void>.delayed(Duration.zero);
    expect(service.dimmed, isTrue);
    fake.emit(testSample(accuracy: HeadingAccuracy.medium));
    await Future<void>.delayed(Duration.zero);
    expect(service.dimmed, isFalse);
    fake.emit(testSample(accuracy: HeadingAccuracy.unreliable));
    await Future<void>.delayed(Duration.zero);
    expect(service.dimmed, isTrue);
  });

  test('stream error surfaces without crashing', () async {
    final fake = FakeHeadingSource();
    final service = supportedService(fake);
    await service.start();
    fake.emitError(StateError('sensor lost'));
    await Future<void>.delayed(Duration.zero);
    expect(service.lastError, isStateError);
  });

  test('status failure surfaces without crashing', () async {
    final fake = FakeHeadingSource()..statusThrow = StateError('no binder');
    final service = supportedService(fake);
    await service.ensureStarted();
    expect(service.lastError, isStateError);
    expect(service.active, isFalse);
  });

  test('dispose cancels the subscription', () async {
    final fake = FakeHeadingSource();
    final service = HeadingService(source: fake);
    await service.start();
    expect(fake.samplesController.hasListener, isTrue);
    service.dispose();
    await Future<void>.delayed(Duration.zero);
    expect(fake.samplesController.hasListener, isFalse);
  });

  test('channel accuracy parsing maps every code', () {
    expect(ChannelHeadingSource.parseAccuracy('high'), HeadingAccuracy.high);
    expect(
      ChannelHeadingSource.parseAccuracy('medium'),
      HeadingAccuracy.medium,
    );
    expect(ChannelHeadingSource.parseAccuracy('low'), HeadingAccuracy.low);
    expect(
      ChannelHeadingSource.parseAccuracy('unreliable'),
      HeadingAccuracy.unreliable,
    );
    expect(
      ChannelHeadingSource.parseAccuracy('bogus'),
      HeadingAccuracy.unreliable,
    );
    expect(ChannelHeadingSource.parseAccuracy(null), HeadingAccuracy.unreliable);
  });

  test('channel sample parsing keeps typed values', () {
    final sample = ChannelHeadingSource.parseSample(const <String, Object?>{
      'magnetic': 270.5,
      'true': 272.0,
      'accuracy': 'medium',
      'timeNanos': 99,
    }, DateTime(2026));
    expect(sample?.magnetic.degrees, 270.5);
    expect(sample?.magnetic.source, 'rotation-vector');
    expect(sample?.trueNorthDeg, 272.0);
    expect(sample?.accuracy, HeadingAccuracy.medium);
    expect(sample?.sensorTimeNanos, 99);
    expect(sample?.receivedAt, DateTime(2026));
  });

  test('channel sample parsing rejects maps without magnetic', () {
    expect(
      ChannelHeadingSource.parseSample(
        const <String, Object?>{'accuracy': 'high'},
        DateTime(2026),
      ),
      isNull,
    );
    expect(ChannelHeadingSource.parseSample(null, DateTime(2026)), isNull);
  });
}

File: apps/atlas-android/app/src/test/java/com/sovereignatlas/atlas/heading/HeadingServiceTest.kt
// ============================================================
// As Above, So Below. As Within, So Without.
// The Future Dictates the Past and the Past is Always Present.
// ============================================================

package com.sovereignatlas.atlas.heading

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

final class FakeHeadingSource(var supported: Boolean = true) : HeadingSource {
    var subscribeCount = 0
    private var sampleListener: ((HeadingSample) -> Unit)? = null

    override fun querySupported(): Boolean = supported

    override fun setSampleListener(listener: ((HeadingSample) -> Unit)?) {
        if (listener != null) subscribeCount++
        sampleListener = listener
    }

    override fun setErrorListener(listener: ((Throwable) -> Unit)?) = Unit

    override fun setDoneListener(listener: (() -> Unit)?) = Unit

    fun emit(sample: HeadingSample) {
        sampleListener?.invoke(sample)
    }
}

private fun sample(
    magnetic: Double = 90.0,
    trueNorth: Double? = 91.0,
    accuracy: HeadingAccuracy = HeadingAccuracy.high,
): HeadingSample {
    return HeadingSample(
        magnetic = AtlasHeading(degrees = magnetic, source = "rotation-vector"),
        accuracy = accuracy,
        receivedAtMs = 1L,
        trueNorthDeg = trueNorth,
    )
}

final class HeadingServiceTest {
    @Test
    fun unsupportedNeverSubscribes() {
        val fake = FakeHeadingSource(supported = false)
        val service = HeadingService(fake)
        service.ensureStarted()
        assertTrue(service.isUnsupported())
        assertEquals(0, fake.subscribeCount)
        assertFalse(service.isActive())
    }

    @Test
    fun supportedStartsAndIngests() {
        val fake = FakeHeadingSource()
        val service = HeadingService(fake)
        service.ensureStarted()
        fake.emit(sample())
        assertEquals(91.0, service.displayDeg())
        assertEquals("TRUE", service.frameLabel())
        assertTrue(service.isActive())
    }

    @Test
    fun magneticFallbackLabelsMag() {
        val fake = FakeHeadingSource()
        val service = HeadingService(fake)
        service.start()
        fake.emit(sample(magnetic = 45.0, trueNorth = null))
        assertEquals(45.0, service.displayDeg())
        assertEquals("MAG", service.frameLabel())
    }

    @Test
    fun dimmingMatrix() {
        val fake = FakeHeadingSource()
        val service = HeadingService(fake)
        service.start()
        fake.emit(sample(accuracy = HeadingAccuracy.high))
        assertFalse(service.isDimmed())
        fake.emit(sample(accuracy = HeadingAccuracy.medium))
        assertFalse(service.isDimmed())
        fake.emit(sample(accuracy = HeadingAccuracy.low))
        assertTrue(service.isDimmed())
        fake.emit(sample(accuracy = HeadingAccuracy.unreliable))
        assertTrue(service.isDimmed())
    }

    @Test
    fun noSampleMeansUnknownAndDimmed() {
        val service = HeadingService(FakeHeadingSource())
        assertEquals(HeadingAccuracy.unknown, service.accuracy())
        assertNull(service.displayDeg())
        assertTrue(service.isDimmed())
    }

    @Test
    fun startIsIdempotentAndStopDetaches() {
        val fake = FakeHeadingSource()
        val service = HeadingService(fake)
        service.start()
        service.start()
        assertEquals(1, fake.subscribeCount)
        service.stop()
        assertFalse(service.isActive())
        fake.emit(sample())
        assertNull(service.displayDeg())
    }

    @Test
    fun headingNormalizes() {
        assertEquals(350.0, AtlasHeading(degrees = -10.0).normalized, 0.0)
        assertEquals(10.0, AtlasHeading(degrees = 370.0).normalized, 0.0)
    }
}
